multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java

```diff
@@ -82,84 +108,114 @@ public ResponseEntity<List<FileMetadata>> getFiles(String spaceGuid, String name
-    private List<FileEntry> uploadFiles(HttpServletRequest request, String spaceGuid, String namespace)
-        throws FileUploadException, IOException, FileStorageException {
-        List<FileEntry> uploadedFiles = new ArrayList<>();
-        FileItemIterator fileItemIterator = createFileIterator(request);
+    @Override
+    public ResponseEntity<Void> startUploadFromUrl(String spaceGuid, String namespace, FileUrl fileUrl) {
+        String decodedUrl = new String(Base64.getUrlDecoder()
+                                             .decode(fileUrl.getFileUrl()));
+        String urlWithoutUserInfo = UriUtil.stripUserInfo(decodedUrl);
+        LOGGER.trace("Received upload from URL {} request", urlWithoutUserInfo);
 
-        while (fileItemIterator.hasNext()) {
-            FileItemStream item = fileItemIterator.next();
-            if (item.isFormField()) {
-                continue; // ignore simple (non-file) form fields
-            }
+        var existingJob = getExistingJob(spaceGuid, namespace, urlWithoutUserInfo);
+        if (existingJob != null) {
+            LOGGER.debug("Async upload job for URL {} exists: {}", urlWithoutUserInfo, existingJob);
+            return ResponseEntity.status(HttpStatus.SEE_OTHER)
+                                 .header("Location", "api/v1/spaces/" + spaceGuid + "/files/jobs/" + existingJob
+                                                                                                            .getId())
+                                 .build();
+        }
 
-            try (InputStream in = item.openStream()) {
-                FileEntry entry;
-                long fileSize = getFileSizeHeaderValue(request);
-                if (fileSize != -1) {
-                    entry = fileService.addFile(spaceGuid, namespace, item.getName(), in, fileSize);
-                } else {
-                    entry = fileService.addFile(spaceGuid, namespace, item.getName(), in);
-                }
-                uploadedFiles.add(entry);
-            }
+        var entry = createJobEntry(spaceGuid, namespace, urlWithoutUserInfo);
+        LOGGER.debug("Creating async upload job for URL {} with ID: {}", urlWithoutUserInfo, entry.getId());
+        uploadJobService.add(entry);
+        try {
+            deployFromUrlExecutor.execute(() -> uploadFileFromUrl(entry, spaceGuid, namespace, decodedUrl));
+        } catch (RejectedExecutionException ignored) {
+            LOGGER.debug("Async upload job {} rejected. Deleting entry", entry.getId());
+            resilientOperationExecutor.execute((Runnable) () -> uploadJobService.createQuery()
+                                                                                .id(entry.getId())
+                                                                                .delete());
+            return ResponseEntity.status(HttpStatus.TOO_MANY_REQUESTS)
+                                 .header(HttpHeaders.RETRY_AFTER, String.valueOf(TimeUnit.MINUTES.toSeconds(5)))
```

Comment:
Can we make it random between 30 seconds and a minute?
Is it possible the CLI to check for this header and retry after this seconds instead of 3 seconds